### ISSUE TITLE
Fully qualified Names of source files

### ISSUE THREAD

**Issue by olbrichj** [xcodeswift/xcproj#40]
Hey,

sorry if this is the wrong place to ask, but I can't find anything anywhere else.

Is there an option to get the fully qualified name of the sourcefiles? In project.pbxproj.objects I can only find the direct filename.

Thanks

**Comment by yonaskolb** [comment]
Do you mean the path to the file? If so you would have to traverse the groups it exists in, and build up the path that way. 
That would be a good helper method, we should add it

**Comment by olbrichj** [comment]
Thanks :)
Yes I'm referring to the path to the file. Either starting at root of the project, or even better at root of the system /Users/xx/project/.../main.swift.
I'll check if I figure out how to traverse it. Currently I'm struggling with the documentation.

**Comment by olbrichj** [comment]
I've written a small example. Not sure if I got everything. Is there an option to separate by target?

let project = try! XcodeProj(path: "Demo.xcodeproj")
let pbxproj = project.pbxproj
let buildGroups = pbxproj.objects.groups

for group in buildGroups {
    if let path = group.path {
        for child in group.children {
            if let found = pbxproj.objects.index(where: { $0.reference == child }) {
                var name = ""
                switch pbxproj.objects[found] {
                case .pbxFileReference(let file): name = file.path!
                default:
                    break
                }
                print(path + "/" + name)
            }
        }
    }
}

**Comment by yonaskolb** [comment]
- instead of `pbxproj.objects.index(where:` you could use `pbxproj.objects.first(where:`
- you'd have to do this recursively to handle nested groups, and build up the path along the way
- you'd have to handle the path differently base on the group `sourceTree`
- if you want to filter by target, you'd have to check if there was a build file for a certain target

What's your use case? Maybe there's a different way to achieve what you're after

**Comment by olbrichj** [comment]
Thanks for your input. 
Currently I'm trying to write some kind of IDE for iOS/OSX development. To get a list of all source files I'd love to import the Xcode.proj. This way I would know, which files are actually used and not just part of the structure.

**Comment by pepibumur** [comment]
This shouldn't be closed. Reopening.

**Comment by ilyapuchka** [comment]
I'm using this code to get full path of file element:

```swift
extension PBXProj.Objects {
    func fullPath(fileElement: PBXFileElement, reference: String, sourceRoot: Path) -> Path? {
        switch fileElement.sourceTree {
        case .absolute?:
            return fileElement.path.flatMap({ Path($0) })
        case .sourceRoot?:
            return fileElement.path.flatMap({ Path($0, relativeTo: sourceRoot) })
        case .group?:
            guard let group = pbxproj.objects.groups.first(where: { $0.value.children.contains(reference) }) else { return sourceRoot }
            guard let groupPath = fullPath(fileElement: group.value, reference: group.key, sourceRoot: sourceRoot) else { return nil }
            return fileElement.path.flatMap({ Path($0, relativeTo: groupPath) })
        default:
            return nil
        }
    }
}
extension Path {
    init(_ string: String, relativeTo relativePath: Path) {
        var path = Path(string)
        if !path.isAbsolute {
            path = (relativePath + path).absolute()
        }
        self.init(path.string)
    }
}
```

It only encounters absolute paths and paths relative to groups or sources root, but it should be possible to extend it to other source tree types if it's possible to provide corresponding environment variables.
P.S. it probably can be optimised as it iterates over all groups for each intermediate group.

**Comment by pepibumur** [comment]
@ilyapuchka I think that would be a great addition to xcproj. Would you mind adding it? We could optimize it in future iterations if it's necessary.

**Comment by olbrichj** [comment]
These methods are great. Just one question:

How to I generate the sourceRoot? Since the project file can be anywhere in subdirectories I'm wondering whether I can get the sourceRoot somewhere within xcproj.

**Comment by ilyapuchka** [comment]
Basically it's just the path to the project file. In Xcode it is an environment variable so you can't get it from the project, unless it defines a custom build setting that equals to this value.

**Comment by ilyapuchka** [comment]
Closed with #213

**Comment by olbrichj** [comment]
That makes it easy :) thx
